dynamic distpatch of fp8 kernels

jeffdaily opened the pull request:
This mostly affects ROCm with hardware that can support one or the other FP8 type. All fp8 kernels are now templated on fp8_type instead of assuming a single fp8_type via `using FP8_TYPE = `. CUDA is largely unaffected.  For CUDA, fp8 kernels only instantiate the OCP type; no binary bloat. For ROCm, two kernel templates are instantiated, one for each fp8 type.

- FP8_TYPE is removed.
- All fp8 kernels have additional fp8_type template param.
- The FP8_E4M3_MAX is replaced with templated struct helper fp8_e4m3_adjusted_max_v.
- is_fp8_ocp() C++ function for host runtime query of preferred fp8 type.
- fp8 kernel launches for ROCm get both OCP and FNUZ templated variants.
- add python APIs current_platform.supports_fp8(), is_fp8_fnuz(), and fp8_dtype().
- add VLLM_DISPATCH_FP8_TYPES that can nest with other dispatch macros

jeffdaily committed:
dynamic distpatch of fp8 kernels

This mostly affects ROCm with hardware that can support one or the other
FP8 type. All fp8 kernels are now templated on fp8_type instead of
assuming a single fp8_type via `using FP8_TYPE = `. CUDA is largely
unaffected.  For CUDA, fp8 kernels only instantiate the OCP type; no
binary bloat. For ROCm, two kernel templates are instantiated, one for
each fp8 type.

- FP8_TYPE is removed.
- All fp8 kernels have additional fp8_type template param.
- The FP8_E4M3_MAX is replaced with templated class FP8_E4M3_ADJUSTED_MAX.
- is_fp8_ocp() C++ function for host runtime query of preferred fp8 type.
- fp8 kernel launches for ROCm get both OCP and FNUZ templated variants.
- add python APIs current_platform.supports_fp8() and is_fp8_fnuz()

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily committed:
Merge branch 'main' into is_fp8_fnuz

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily committed:
lint

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

ProExpertProg commented on benchmarks/kernels/benchmark_moe.py:
```diff
@@ -17,7 +17,7 @@
 from vllm.utils import FlexibleArgumentParser
 
-FP8_DTYPE = torch.float8_e4m3fnuz if current_platform.is_rocm(
+FP8_DTYPE = torch.float8_e4m3fnuz if current_platform.is_fp8_fnuz(
 ) else torch.float8_e4m3fn
```
This appears in many places - could we just extract it into `current_platform.fp8_dtype()`?

jeffdaily replied:
Great idea.  But I will also keep `is_fp8_fnuz()` because it is a useful shorthand before calling `normalize_e4m3fn_to_e4m3fnuz`.

ProExpertProg replied:
Yep, agreed 😃

ProExpertProg commented on csrc/dispatch_utils.h:
```diff
@@ -14,15 +14,18 @@
+    AT_DISPATCH_CASE(at::ScalarType::Float8_e4m3fn, __VA_ARGS__)   \
     AT_DISPATCH_CASE(at::ScalarType::Float8_e4m3fnuz, __VA_ARGS__) \
     AT_DISPATCH_CASE(at::ScalarType::Char, __VA_ARGS__)
+#else
```
We could add a `#define VLLM_DISPATCH_CASE_FP8_TYPES` here as well, with only one case on CUDA and 2 cases on ROCm

jeffdaily replied:
I could, but where/how do you see that being used?  

We already have `VLLM_DISPATCH_CASE_QUANT_TYPES(...)` where I added the two fp8 cases on rocm while cuda still has just the one.

Then in [csrc/layernorm_quant_kernels.cu](https://github.com/vllm-project/vllm/blob/04222984f870f3e4ce92076a322e55b18427f40a/csrc/layernorm_quant_kernels.cu#L179-L185), the dispatch is based on the `input.scalar_type()` but we need to also conditionalize the dispatch based on the `out.scalar_type()`.  We're not dispatching based on the fp8 type.  

Did you mean something else?

ProExpertProg replied:
Yes, I imagined nesting the dispatching macros. I've seen it before but we can definitely keep it this way if nesting doesn't work.

jeffdaily replied:
Nesting dispatch macros was the way to go.

ProExpertProg commented on csrc/layernorm_quant_kernels.cu:
```diff
@@ -176,21 +176,35 @@ void rms_norm_static_fp8_quant(torch::Tensor& out,     // [..., hidden_size]
-            weight.data_ptr<scalar_t>(), scale.data_ptr<float>(), epsilon,
-            num_tokens, hidden_size);
-  });
+  if (is_fp8_ocp()) {
```
use new dispatch macro here

ProExpertProg commented on csrc/layernorm_quant_kernels.cu:
```diff
@@ -227,8 +241,22 @@ void fused_add_rms_norm_static_fp8_quant(
   if (ptrs_are_aligned && hidden_size % 8 == 0) {
-    LAUNCH_FUSED_ADD_RMS_NORM(8);
+    if (is_fp8_ocp()) {
+      LAUNCH_FUSED_ADD_RMS_NORM(8, c10::Float8_e4m3fn);
+    }
+#ifdef USE_ROCM
+    else {
+      LAUNCH_FUSED_ADD_RMS_NORM(8, c10::Float8_e4m3fnuz);
+    }
+#endif
   } else {
-    LAUNCH_FUSED_ADD_RMS_NORM(0);
+    if (is_fp8_ocp()) {
+      LAUNCH_FUSED_ADD_RMS_NORM(0, c10::Float8_e4m3fn);
+    }
+#ifdef USE_ROCM
+    else {
+      LAUNCH_FUSED_ADD_RMS_NORM(0, c10::Float8_e4m3fnuz);
+    }
+#endif
```
Could clean up dispatching logic here

jeffdaily replied:
The nested dispatch really helped clean this up.

ProExpertProg commented on csrc/quantization/fp8/common.cuh:
```diff
@@ -7,18 +7,47 @@
+}
+
+template <typename T>
+class FP8_E4M3_ADJUSTED_MAX {
```
I think this should be an empty `struct`, that's more idiomatic C++, and you can avoid the explicit `public:`. Also not all upper case, that implies a constant and not class (struct) name.

jeffdaily replied:
Done.

ProExpertProg commented on csrc/quantization/fp8/common.cuh:
```diff
@@ -7,18 +7,47 @@
+  static constexpr c10::Float8_e4m3fnuz val() {
+    return c10::Float8_e4m3fnuz(0x7E, c10::Float8_e4m3fnuz::from_bits());
+  }
+};
```
Add a "value" helper here, like
```cpp
template <typename T>
static constexpr T fp8_e4m3_adjusted_max_v = fp8_e4m3_adjusted_max<T>::value();
```

jeffdaily replied:
Done.

ProExpertProg commented on csrc/quantization/fp8/common.cuh:
```diff
@@ -32,25 +61,35 @@ __device__ __forceinline__ float atomicMaxFloat(float* addr, float value) {
   }
 
-  float r = fmax(-FP8_E4M3_MAX, fmin(x, FP8_E4M3_MAX));
+  float r = fmax(-FP8_E4M3_ADJUSTED_MAX<fp8_type>::val(),
```
`fp8_e4m3_adjusted_max_v<fp8_type>`

ProExpertProg commented on csrc/quantization/fp8/common.cuh:
```diff
@@ -32,25 +61,35 @@ __device__ __forceinline__ float atomicMaxFloat(float* addr, float value) {
-      __hip_cvt_float_to_fp8(r, fp8::fp8_type::__default_saturation,
-                             fp8::fp8_type::__default_interpret),
-      c10::Float8_e4m3fnuz::from_bits());
+  if constexpr (std::is_same_v<fp8_type, c10::Float8_e4m3fn>) {
+    ret =
+        fp8_type(__hip_cvt_float_to_fp8(r, __hip_fp8_e4m3::__default_saturation,
+                                        __hip_fp8_e4m3::__default_interpret),
+                 fp8_type::from_bits());
+  }
+  if constexpr (std::is_same_v<fp8_type, c10::Float8_e4m3fnuz>) {
+    ret = fp8_type(
+        __hip_cvt_float_to_fp8(r, __hip_fp8_e4m3_fnuz::__default_saturation,
+                               __hip_fp8_e4m3_fnuz::__default_interpret),
+        fp8_type::from_bits());
+  }
```
Should this be extracted to a template-specialized utility functions for possible future reuse?

jeffdaily replied:
Done.

ProExpertProg commented on csrc/quantization/fp8/common.cuh:
```diff
@@ -32,25 +61,35 @@ __device__ __forceinline__ float atomicMaxFloat(float* addr, float value) {
-__device__ __forceinline__ FP8_TYPE scaled_fp8_conversion(float const val,
-                                                          float const scale) {
+template <bool is_scale_inverted, typename fp8_type>
+__device__ __forceinline__ void scaled_fp8_conversion(fp8_type& ret,
```
Why the update by reference instead of returning the value?

jeffdaily replied:
I added `fp8_type` as a template param and I panicked that you can't have two functions that differ in just the return type in C++.

jeffdaily replied:
Turns out this works just fine the way it was before and with fp8_type as a template param.  Will revert.

ProExpertProg commented on csrc/quantization/fused_kernels/quant_conversions.cuh:
```diff
@@ -54,16 +57,34 @@ struct ScaledQuant<
-struct ScaledQuant<
-    quant_type_t, is_scale_inverted,
-    typename std::enable_if_t<std::is_same_v<quant_type_t, FP8_TYPE>>> {
+struct ScaledQuant<quant_type_t, is_scale_inverted,
+                   typename std::enable_if_t<
+                       std::is_same_v<quant_type_t, c10::Float8_e4m3fn>>> {
+  static __device__ __forceinline__ quant_type_t quant_fn(float const x,
+                                                          float const scale) {
+    quant_type_t ret;
+    if constexpr (is_scale_inverted) {
+      float_to_fp8(ret, x * scale);
+    } else {
+      float_to_fp8(ret, x / scale);
+    }
+    return ret;
+  }
+};
+
+template <typename quant_type_t, bool is_scale_inverted>
+struct ScaledQuant<quant_type_t, is_scale_inverted,
+                   typename std::enable_if_t<
+                       std::is_same_v<quant_type_t, c10::Float8_e4m3fnuz>>> {
```
I think theses could just stay the same struct and just expand the condition:

```suggestion
struct ScaledQuant<quant_type_t, is_scale_inverted,
                   typename std::enable_if_t<
                       std::is_same_v<quant_type_t, c10::Float8_e4m3fn> || 
                       std::is_same_v<quant_type_t, c10::Float8_e4m3fnuz>>> {
```

ProExpertProg commented on csrc/quantization/fused_kernels/quant_conversions.cuh:
```diff
@@ -31,9 +31,12 @@ static __device__ __forceinline__ int8_t float_to_int8_rn(float const x) {
-  float const r = fmax(-FP8_E4M3_MAX, fmin(x, FP8_E4M3_MAX));
-  return static_cast<FP8_TYPE>(r);
+template <typename fp8_type>
+static __device__ __forceinline__ void float_to_fp8(fp8_type& ret,
```
Same here, curious why you changed this to update by reference?

jeffdaily replied:
Reverted.

ProExpertProg commented on tests/kernels/test_triton_scaled_mm.py:
```diff
@@ -32,11 +32,11 @@ def scaled_mm_torch(a: torch.Tensor,
-    elif current_platform.is_cuda() and supports_fp8:
-        types.append(torch.float8_e4m3fn)
+    if current_platform.supports_fp8:
+        if current_platform.is_fp8_fnuz():
+            types.append(torch.float8_e4m3fnuz)
+        else:
+            types.append(torch.float8_e4m3fn)
```
Nice simplification! I think `current_platform.fp8_type` would simplify this further

jeffdaily replied:
Agree. Done.

ProExpertProg commented on csrc/quantization/fp8/common.cuh:
```diff
@@ -7,18 +7,47 @@
 #endif
-constexpr static auto kFp8Type = c10::CppTypeToScalarType<FP8_TYPE>::value;
+
+static bool is_fp8_ocp() {
```
Could you add a brief comment about this method? I'd mention it checks device properties and .

It would also be good to check the property directly instead of querying all device properties, which is significantly slower (in my experience on CUDA, not sure if the same applies on AMD)

jeffdaily replied:
I will add a comment, yes.

Would it ease your mind to know that `at::cuda::getCurrentDeviceProperties()` is cached internally by pytorch?

See https://github.com/pytorch/pytorch/blob/ed9c8a5d136cae1dac46c7700b83d3e07830cff8/aten/src/ATen/cuda/CUDAContext.cpp#L17-L60 for the implementation.

ProExpertProg reviewed (changes requested):
Nice PR! Thanks for removing TODOs and cleaning up a variety of nested if statements etc.

I left a few comments for further improving the code quality - this is a pretty messy part of our codebase and we should really keep it from getting more convoluted - thanks for doing your part already!

jeffdaily committed:
add current_platform.fp8_dtype()

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily committed:
requested changes

- FP8_E4M3_ADJUSTED_MAX -> fp8_e4m3_adjusted_max_v
- struct fp8_e4m3_adjusted_max
- ScaledQuant collapse duplicate fp8 definition

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily committed:
more suggested changes

- added template-specialized utility function fp8::cvt_c10
- revert update by reference back to returning value

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily committed:
nested dispatch

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily commented:
@ProExpertProg I really appreciate your thoughtful review.  It greatly improved the quality.

ProExpertProg commented on csrc/dispatch_utils.h:
```diff
@@ -6,6 +6,10 @@
 
+// need a special dispatch case macro since we will nest the FP8 dispatch
+#define AT_DISPATCH_FP8_CASE(enum_type, ...) \
+  AT_PRIVATE_CASE_TYPE_USING_HINT(enum_type, fp8_t, __VA_ARGS__)
```
This is nice!

ProExpertProg commented on csrc/dispatch_utils.h:
```diff
@@ -14,17 +18,30 @@
+    AT_DISPATCH_CASE(at::ScalarType::Char, __VA_ARGS__)
 #endif
 
+#define VLLM_DISPATCH_FP8_TYPES(TYPE, NAME, ...) \
```
Maybe a quick comment above this macro that mentions that the dispatched type is available as `fp8_t` and not `scalar_t`?

ProExpertProg commented on csrc/quantization/fp8/common.cuh:
```diff
@@ -7,18 +7,48 @@
+};
+
+// Using the default max value from pytorch (240.0 0x7F) will cause accuracy
+// issue when running dynamic quantization. Here use 224.0 0x7E for rocm.
```
Nit: typo (issues)

ProExpertProg commented on vllm/model_executor/layers/quantization/utils/fp8_utils.py:
```diff
@@ -150,8 +146,7 @@ def input_to_float8(
     "with tensor-wise quantization."""
     if dtype is None:
-        dtype = (torch.float8_e4m3fnuz
-                 if current_platform.is_rocm() else torch.float8_e4m3fn)
+        dtype = current_platform.fp8_dtype()
```
Nit: could be `dtype = current_platform.fp8_dtype() if dtype is None else dtype` if it fits in a line

ProExpertProg commented on vllm/platforms/interface.py:
```diff
@@ -330,6 +330,27 @@ def get_device_communicator_cls(cls) -> str:
+    @classmethod
+    def is_fp8_fnuz(cls) -> bool:
+        """
+        Returns whether the preferred F8 type is fnuz on the current platform.
```
Maybe add a comment here about what platforms use this dtype?

jeffdaily replied:
Done.  I also added links to the 2 different FP8 standards for additional clarification, and the other API `fp8_dtype` refers back to the doc for `is_fp8_fnuz` to avoid copy/paste.

ProExpertProg reviewed (approved):
Thanks for addressing the comments! It looks great now.

Added couple more nits, and I think it would be nice to add a comment on the C++ side somewhere that explains there are two FP8 types on ROCm (not sure where the best place for that is). Otherwise LGTM!

tlrmchlsmth commented on vllm/platforms/interface.py:
```diff
@@ -330,6 +330,27 @@ def get_device_communicator_cls(cls) -> str:
         return "vllm.distributed.device_communicators.base_device_communicator.DeviceCommunicatorBase"  # noqa
 
+    @classmethod
+    def supports_fp8(cls) -> bool:
```
If we add this, we’ll need to audit other platforms — TPUs at least should support fp8 (cc @robertgshaw2-redhat @mgoin )

jeffdaily replied:
Does the audit need to happen as part of this PR or can it be a follow-up PR?

`supports_fp8` was added to replace `current_platform.has_device_capability(89)` which was incorrect on ROCm.  It was only used in tests:

- tests/quantization/test_fp8.py
- tests/kernels/test_cutlass.py
- tests/kernels/test_triton_scaled_mm.py

There should be a follow up task to audit all call sites of `has_device_capability` because those are checks for platform features and it would be better to have specific APIs for them in `current_platform`.  For example, in [gpu_worker.py](https://github.com/vllm-project/vllm/blob/bc2d4473bf943feb8abac1688dbc7c5701d7282d/vllm/v1/worker/gpu_worker.py#L276) it's used to check for bfloat16 support.  The code would improve if there was a `current_platform.supports_bfloat16()`.

mgoin replied:
TPUs do not support fp8 yet. I think we only need to consider CUDA and ROCm at the moment

ProExpertProg replied:
@jeffdaily would you mind creating a GitHub issue for the device capability audit you're describing? I agree it would be a great idea and agreed it is out of the scope of this PR.

jeffdaily replied:
https://github.com/vllm-project/vllm/issues/14637

jeffdaily committed:
additional reviewer requests

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily commented:
@ProExpertProg Getting some cuda build failures:

```
error: identifier "fp8_e4m3_adjusted_max_v< ::c10::Float8_e4m3fn> " is undefined in device code
```

Looks like fp8_e4m3_adjusted_max_v is the problem, at least for nvcc.  Any suggestions?

ProExpertProg commented:
Maybe add `C10_HOST_DEVICE` back in, like the original constant?

jeffdaily committed:
conditionally add back C10_HOST_DEVICE for fp8 max

try to fix fp8_e4m3_adjusted_max_v is undefined in device code on cuda

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily committed:
revert removal of current_platform_fp8_dtype

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily commented:
@ProExpertProg fixed.  

I had aggressively removed that variable from vllm/model_executor/layers/quantization/utils/fp8_utils.py because it wasn't used elsewhere in that file.  I failed to grep through all the other files for anyone that might have imported it.

ProExpertProg commented:
Yeah that was nice cleanup! @robertgshaw2-redhat @mgoin could you guys help with the fp8 support on other platforms?

ProExpertProg commented on vllm/model_executor/layers/quantization/utils/fp8_utils.py:
```diff
@@ -21,9 +21,7 @@
-current_platform_fp8_dtype = (torch.float8_e4m3fnuz
-                              if current_platform.is_rocm() else
-                              torch.float8_e4m3fn)
+current_platform_fp8_dtype = current_platform.fp8_dtype()
```
I think it would actually be better to get rid of this and use `current_platform.fp8_dtype()` directly

jeffdaily replied:
Ok, done. Removed it again and updated corresponding code.

jeffdaily committed:
remove current_platform_fp8_dtype

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

jeffdaily committed:
Merge branch 'main' into is_fp8_fnuz

ProExpertProg reviewed (approved):
I think this should be good to go!

robertgshaw2-redhat reviewed (approved):
This is an excellent PR

ProExpertProg commented:
@jeffdaily I am planning to look into it more but do you know if this is a ROCm version issue?
```
/mnt/nvme3n1p1/sage/git/nm-vllm/csrc/quantization/fp8/amd/quant_utils.cuh:25:33: error: use of undeclared identifier '__hip_fp8_e4m3'
   25 |       __hip_cvt_float_to_fp8(r, __hip_fp8_e4m3::__default_saturation,
      |                                 ^
/mnt/nvme3n1p1/sage/git/nm-vllm/csrc/quantization/fp8/amd/quant_utils.cuh:26:30: error: use of undeclared identifier '__hip_fp8_e4m3'
   26 |                              __hip_fp8_e4m3::__default_interpret),
      |                              ^
```

jeffdaily commented:
Looks like a ROCm version issue.  `__hip_fp8_e4m3` and related types were added in ROCm 6.3.  Not present in ROCm 6.2.

ProExpertProg commented:
Could you help with the fix? I assume there's an alternative for ROCm 6.2?

jeffdaily commented:
Working on a fix now.  Can we forward-fix with a new PR or do you need to revert this one?

ProExpertProg commented:
I think forward fixing is fine, the revert wouldn't be immediate either anyway. Unless you think the fix will be complicated?

jeffdaily commented:
Fix shouldn't be complicated.  Will need to isolate the use of `__hip_fp8_e4m3` behind a ROCM_VERSION macro limiting it to ROCm 6.3 or newer.

ProExpertProg commented:
Okay ping me on vLLM Slack once done so we can merge ASAP

jeffdaily commented:
@ProExpertProg I'm not on slack yet.  Here's the PR fix.

https://github.com/vllm-project/vllm/pull/14709
